### PR TITLE
Orion XS support

### PR DESCRIPTION
**Pull request by blammit**
This is a big PR, but hopefully the commit separation helps.

### CODE REVIEW

**Comment by DanielMcInnes** [comment]
Not sure if this is part of these changes

Screenshot from 2024-03-12 16-41-58

File: pages/OverviewPage.qml [hunk 1]
```diff
@@ -327,10 +339,9 @@ SwipeViewPage {
 			return VenusOS.WidgetConnector_AnimationMode_NotAnimated
 		}
 		// Assumes startWidget is the AC/DC input widget.
```
Comment (jpetrell) [inline pages/OverviewPage.qml:341]: Is this comment still valid? ^

File: pages/OverviewPage.qml [hunk 2]
```diff
@@ -327,10 +339,9 @@ SwipeViewPage {
...
-		if (!connectorWidget.startWidget.input) {
-			return VenusOS.WidgetConnector_AnimationMode_NotAnimated
-		}
-		const power = connectorWidget.startWidget.input.power
+		// Use the displayed power to calculate whether the connector should be animated.
+		const power = !connectorWidget.startWidget.quantityLabel.dataObject ? NaN
+				: connectorWidget.startWidget.quantityLabel.dataObject.power
```
Comment (jpetrell) [inline pages/OverviewPage.qml:344]: Should this be a binding?

File: components/QuantityGroupListHeader.qml [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+		id: firstTitleLabel
+		anchors {
+			bottom: parent.bottom
+			bottomMargin: Theme.geometry_quantityTableSummary_verticalMargin
+		}
+		leftPadding: Theme.geometry_listItem_content_horizontalMargin
+		font.pixelSize: Theme.font_size_caption
+		color: Theme.color_solarListPage_header_text
+		elide: Text.ElideRight
+		width: parent.width - quantityRow.width
```
Comment (jpetrell) [inline components/QuantityGroupListHeader.qml:27]: Add tiny spacing in between in case the text overflows?

File: pages/solar/SolarDeviceListPage.qml [hunk 4]
```diff
@@ -108,19 +114,18 @@ Page {
...
 				}
+				leftPadding: Theme.geometry_listItem_content_horizontalMargin
 				text: chargerMode
 						//% "PV Charger"
 					  ? qsTrId("solardevices_pv_charger")
 					  : CommonWords.pv_inverter
 				font.pixelSize: Theme.font_size_caption
 				color: Theme.color_solarListPage_header_text
 				elide: Text.ElideRight
+				width: parent.width - quantityRow.width
```
Comment (jpetrell) [inline pages/solar/SolarDeviceListPage.qml]: Same as before, maybe add small spacing in case of overflow.

File: data/mock/DcInputsImpl.qml [hunk 5]
```diff
@@ -9,16 +9,17 @@ import Victron.VenusOS
 QtObject {
 	id: root
 
+	property int mockDeviceCount
+	property var _createdObjects: []
+
 	function populate() {
-		// Add a random set of DC inputs.
+		 // Add a random set of DC inputs.
```
Comment (jpetrell) [inline data/mock/DcInputsImpl.qml:16]: Unnecessary whitespace?

File: pages/settings/devicelist/dc-in/PageAlternatorModel.qml [hunk 6]
```diff
@@ -11,32 +11,32 @@ ObjectModel {
...
-		}
-		VeQuickItem {
-			id: dcCurrent
-			uid: root.bindPrefix + "/Dc/0/Current"
-		}
-		VeQuickItem {
-			id: dcPower
-			uid: root.bindPrefix + "/Dc/0/Power"
-		}
+	ListDcInputMeasurements {
```
Comment (jpetrell) [inline pages/settings/devicelist/dc-in/PageAlternatorModel.qml:22]: Kind of wish there were more in the name that hinted this is ListQuantityGroup, e.g. why talk about measurements and not quantities. ListDcInputQuantityGroup?

**Comment by jpetrell** [comment]
Expecting the quantity group '|' separators to horizontally align between the subsequent list items, but I guess not trivial to achieve and existing issue not caused by this PR:
image

File: pages/settings/devicelist/dc-in/ListCycleHistoryItem.qml [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+
+	ListLabel {
+		text: cycle == 0
+				//% "Active cycle"
+			  ? qsTrId("cycle_history_active")
+				//: %1 = cycle number
+				//% "Cycle %1"
+			  : qsTrId("cycle_history_num").arg(cycle)
+		horizontalAlignment: Text.AlignHCenter
+	}
```
Comment (jpetrell) [inline pages/settings/devicelist/dc-in/ListCycleHistoryItem.qml:23]: Most ListLabels left align, this feels bit unusual, kind of used as secondary header. Also wonder if it is clear to all users what "Active cycle" means.

**Comment by jpetrell** [comment]
The code changes looks good :heavy_check_mark: , but I need a bit of help verifying the actual functionality.

Some observations, which may or may not be valid:
- Using the mock backend ("Multiple alternators, including Orion XS") didn't find a device page that would match Orion XS alternator settings, the DC device with new Overall history and Charge cycle history lacks DC input measurements and network status info, and displays a temperature, which gui v1 example doesn't.
- Similarly using the mock backend didn't find a device page that would match Orion XS .dcdc settings, the DC Load (dcdc) opens the new PageDcDcConverter page, but lacks the switch and DC input quantity group.
- DC Load (dcdc) can be accessed from the Overview page, but not from Device list settings like other DC Load items
- For some DC devices (alternator) the Overall history page is fully empty page. Not sure if this is mock backend issue, but if this can happen in real world scenario could display some empty state placeholder

File: pages/solar/SolarDeviceListPage.qml [hunk 8]
```diff
@@ -76,17 +79,20 @@ Page {
...
-					current: pvInverter.current
-					power: pvInverter.power
-					voltage: pvInverter.voltage
+					quantityRowWidth: Theme.geometry_solarListPage_quantityRow_width
+					quantityModel: [
+						{ value: pvInverter.energy, unit: VenusOS.Units_Energy_KiloWattHour },
+						{ value: pvInverter.voltage, unit: VenusOS.Units_Volt },
+						{ value: pvInverter.current, unit: VenusOS.Units_Amp },
+						{ value: pvInverter.power, unit: VenusOS.Units_Watt },
+					]
```
Comment (jpetrell) [inline pages/solar/SolarDeviceListPage.qml:95]: Getting some warnings, but not caused by your changes here:

```
qrc:/qt/qml/Victron/VenusOS/pages/solar/SolarDeviceListPage.qml:91: TypeError: Cannot read property 'energy' of null
qrc:/qt/qml/Victron/VenusOS/pages/solar/SolarDeviceListPage.qml:88: TypeError: Cannot read property 'name' of null
```

**Comment by blammit** [comment]
@jpetrell This can be tested on device using the recordings supplied by Thiemo in https://github.com/victronenergy/gui-v2/issues/738.

E.g. extract the `OrionXS.zip` to get `OrionXS.rb`, copy it to device, and run:

`/opt/victronenergy/dbus-recorder/play.py <path/to/OrionXS.rb>`

File: pages/settings/devicelist/dc-in/ListCycleHistoryItem.qml [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+
+	ListTextItem {
+		text: CommonWords.status
+		dataItem.uid: root.bindPrefix + "/TerminationReason"
+		secondaryText: {
+			if (dataItem.value === undefined) {
```
Comment (DanielMcInnes) [inline pages/settings/devicelist/dc-in/ListCycleHistoryItem.qml:29]: I think you can have `case undefined` in the following switch statement

File: pages/settings/devicelist/dc-in/PageDcDcConverter.qml [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+** See LICENSE.txt for license information.
+*/
+
+import QtQuick
+import Victron.VenusOS
+
+Page {
+	id: root
+
+	property string bindPrefix: bindPrefix
```
Comment (DanielMcInnes) [inline pages/settings/devicelist/dc-in/PageDcDcConverter.qml:12]: this doesn't look right. Not sure what it means to initialize bindPrefix to itself

**Review by DanielMcInnes — APPROVED** [review]
lgtm, nits

**Comment by blammit** [comment]
> The code changes looks good ✔️ , but I need a bit of help verifying the actual functionality.
> 
> Some observations, which may or may not be valid:
> 
> * Using the mock backend ("Multiple alternators, including Orion XS") didn't find a device page that would match Orion XS alternator settings, the DC device with new Overall history and Charge cycle history lacks DC input measurements and network status info, and displays a temperature, which gui v1 example doesn't.
> * Similarly using the mock backend didn't find a device page that would match Orion XS .dcdc settings, the DC Load (dcdc) opens the new PageDcDcConverter page, but lacks the switch and DC input quantity group.
> * DC Load (dcdc) can be accessed from the Overview page, but not from Device list settings like other DC Load items
> * For some DC devices (alternator) the Overall history page is fully empty page. Not sure if this is mock backend issue, but if this can happen in real world scenario could display some empty state placeholder

I'll update the mock data to reflect some of the recording data from the task. But, there is something odd going on -- the dcdc should definitely be accessible from the Device List, and the Alternator too. Not sure what is happening here.

It's a good question about the empty states -- I don't know if they would ever occur in the real world scenario. But we could show a placeholder just to be safe.

**Comment by blammit** [comment]
> Expecting the quantity group '|' separators to horizontally align between the subsequent list items, but I guess not trivial to achieve and existing issue not caused by this PR: image

Agreed, would be really nice to have that, and it's more easily achievable now that QuantityTableMetrics can be used to provide a consistent width across different list items. Although, if different values have different precisions, that would make things more complicated.

**Comment by blammit** [comment]
> I'll update the mock data to reflect some of the recording data from the task. But, there is something odd going on -- the dcdc should definitely be accessible from the Device List, and the Alternator too. Not sure what is happening here.

Sorry, my mistake -- I hadn't added the `dcdc` devices to the Device List, thanks for catching that!

> * Using the mock backend ("Multiple alternators, including Orion XS") didn't find a device page that would match Orion XS alternator settings, the DC device with new Overall history and Charge cycle history lacks DC input measurements and network status info, and displays a temperature, which gui v1 example doesn't.

I can't find any DC input measurements and network status in the Charge Cycle history in gui-v1, only in the main Alternator details. Can you point me to that?

> * Similarly using the mock backend didn't find a device page that would match Orion XS .dcdc settings, the DC Load (dcdc) opens the new PageDcDcConverter page, but lacks the switch and DC input quantity group.

Will update the mock data with these details.

> * DC Load (dcdc) can be accessed from the Overview page, but not from Device list settings like other DC Load items

Fixed.

> * For some DC devices (alternator) the Overall history page is fully empty page. Not sure if this is mock backend issue, but if this can happen in real world scenario could display some empty state placeholder

Could add some placeholder text here.

File: pages/OverviewPage.qml (hunk 1 above)
Comment (blammit) [reply pages/OverviewPage.qml:341]: Yes

File: pages/OverviewPage.qml (hunk 2 above)
Comment (blammit) [reply pages/OverviewPage.qml:344]: It already functions as a binding, since the function is called from a property binding. Or do you mean a `Binding` object?

File: components/QuantityGroupListHeader.qml (hunk 3 above)
Comment (blammit) [reply components/QuantityGroupListHeader.qml:27]: Done

File: data/mock/DcInputsImpl.qml (hunk 5 above)
Comment (blammit) [reply data/mock/DcInputsImpl.qml:16]: Fixed

File: pages/solar/SolarDeviceListPage.qml (hunk 4 above)
Comment (blammit) [reply pages/solar/SolarDeviceListPage.qml]: Done

File: pages/settings/devicelist/dc-in/PageAlternatorModel.qml (hunk 6 above)
Comment (blammit) [reply pages/settings/devicelist/dc-in/PageAlternatorModel.qml:22]: It's because it refers to a common set of measurements. But maybe QuantityGroup is more useful in this context, have renamed this.

File: pages/settings/devicelist/dc-in/ListCycleHistoryItem.qml (hunk 7 above)
Comment (blammit) [reply pages/settings/devicelist/dc-in/ListCycleHistoryItem.qml:23]: Yes maybe left alignment is nicer? Updated.

Comment (blammit) [reply pages/settings/devicelist/dc-in/ListCycleHistoryItem.qml:23]: I don't know about the "Active cycle", but I assume it's meant to be straightforward to the audience.

File: pages/OverviewPage.qml [hunk 11]
```diff
@@ -258,30 +258,27 @@ SwipeViewPage {
...
+			if (widgetType >= 0) {
+				widget = _createWidget(widgetType)
+				widget.inputs.clear()
 			}
+		}
+		for (i = 0; i < Global.dcInputs.model.count; ++i) {
+			// Add the input to the DC widget
+			const dcInput = Global.dcInputs.model.deviceAt(i)
+			widgetType = _dcWidgetForInputType(dcInput.inputType)
 			if (widgetType < 0) {
```
[the remark this answers is not in the document]
Comment (blammit) [reply pages/OverviewPage.qml]: Fixed

**Comment by mpvader** [comment]
Hey @blammit ,

> For some DC devices (alternator) the Overall history page is fully empty page. Not sure if this is mock backend issue, but if this can happen in real world scenario could display some empty state placeholder
> Could add some placeholder text here.
>

Rather then showing an empty submenu, please hide it.

File: pages/OverviewPage.qml [hunk 12]
```diff
@@ -292,6 +289,21 @@ SwipeViewPage {
 		_leftWidgets = widgetCandidates
 	}
 
+	function _dcWidgetForInputType(dcInputType) {
```
[the remark this answers is not in the document]
Comment (blammit) [reply pages/OverviewPage.qml]: Fixed

File: components/QuantityRepeater.qml [hunk 13]
```diff
@@ -12,9 +12,26 @@ Repeater {
 	property int minimumDelegateWidth
 
 	delegate: Row {
+		// Visibility is determined by optional 'visible' property.
+		readonly property bool showValue: modelData.visible !== false
```
[the remark this answers is not in the document]
Comment (blammit) [reply components/QuantityRepeater.qml:16]: Yes, just wanted to make it more explicit

File: pages/OverviewPage.qml [hunk 14]
```diff
@@ -258,30 +258,27 @@ SwipeViewPage {
...
-				widgetType = VenusOS.OverviewWidget_Type_DcGenerator
-				break
+			// Clear the inputs list for this DC input type
+			widgetType = _dcWidgetForInputType(Global.dcInputs.model.deviceAt(i).inputType)
+			if (widgetType >= 0) {
+				widget = _createWidget(widgetType)
+				widget.inputs.clear()
 			}
+		}
+		for (i = 0; i < Global.dcInputs.model.count; ++i) {
```
[the remark this answers is not in the document]
Comment (blammit) [reply pages/OverviewPage.qml:261]: Done

File: pages/settings/devicelist/dc-in/PageDcDcConverter.qml (hunk 10 above)
Comment (blammit) [reply pages/settings/devicelist/dc-in/PageDcDcConverter.qml]: Strange indeed. Removed

File: pages/settings/devicelist/dc-in/ListCycleHistoryItem.qml (hunk 9 above)
Comment (blammit) [reply pages/settings/devicelist/dc-in/ListCycleHistoryItem.qml]: Done

File: pages/settings/devicelist/dc-in/ListCycleHistoryItem.qml (hunk 7 above)
Comment (jpetrell) [reply pages/settings/devicelist/dc-in/ListCycleHistoryItem.qml:22]: > Yes maybe left alignment is nicer? Updated.

Was discussing separating related items using section headers with Serj the other day and he mentioned there actually being such component already, in use in the Notifications page. Where yeah, the label is left aligned. :+1:

**Comment by jpetrell** [comment]
> extract the OrionXS.zip to get OrionXS.rb, copy it to device, and run:
> /opt/victronenergy/dbus-recorder/play.py <path/to/OrionXS.rb>

That is so cool. And just works with no setup needed. :ok_hand: 

Few findings testing with the Orion recordings:

- For the Orion XS alternator can see two State items, one showing state "Bulk charging", other "Standalone". Looking at the code the first displays /State and the second /Link/NetworkStatus. On SolarChargerNetworkedOperationPage /Link/NetworkStatus list item displays "Network status".

- Error item showing "#0 No error" feels techy. Generator start/stop Error item shows "No error".

>> For some DC devices (alternator) the Overall history page is fully empty page.
>> Could add some placeholder text here.
> Rather then showing an empty submenu, please hide it.

The empty Overall history page could still be seen on the device when running Boat/Motorhome demo 2. If h

**Comment by blammit** [comment]
In the process of fixing the empty "Overall" page and trying to generalize the solution across other pages like the battery settings, I encountered https://github.com/victronenergy/gui-v2/issues/860, so will fix this issue there.

**Comment by mpvader** [comment]
> Error item showing "#0 No error" feels techy. Generator start/stop Error item shows "No error".
> AFAIK this is intentional, to show the specific error code. This comes from the underlying veutil Charger/Alternator errors.

but no need to show #0 if there is no error (!)

the reason to show a code is to make it less ambiguous, only relevant for when there is an error

**Review by jpetrell — APPROVED** [review]
Other than the error code LGTM. Approving already. Need to rebase on top of the latest main to fix the CI failure.

**Comment by jpetrell** [comment]
> Other than the error code LGTM. Approving already. Need to rebase on top of the latest main to fix the CI failure.

Spotted "#0 No errors" in other device settings pages. Should be fixed consistently, so filed a separate task https://github.com/victronenergy/gui-v2/issues/864.

> Need to rebase on top of the latest main to fix the CI failure.

Well not really, as the main already has the CI fix. Merging.
